File: python/cspguardian_py/scanner.py
from __future__ import annotations

import base64
import hashlib
import os
import re
import secrets
from pathlib import Path

from .models import (
    CLEANUP_EXTERNALIZE,
    CLEANUP_HASH,
    CLEANUP_NONCE,
    FRAMEWORK_JS_LEGACY,
    FRAMEWORK_JS_MODERN,
    FRAMEWORK_LEGACY_DOTNET,
    ScanOptions,
    ScanResult,
    Violation,
    VIOLATION_DYNAMIC_INLINE,
    VIOLATION_EVENT_HANDLER,
    VIOLATION_INLINE_SCRIPT,
    VIOLATION_INLINE_STYLE,
    VIOLATION_JAVASCRIPT_URL,
    VIOLATION_LEGACY_RESOURCE,
    VIOLATION_STYLE_ATTRIBUTE,
    VIOLATION_VIEWSTATE_SCRIPT,
)

SUPPORTED_EXTENSIONS = {".html", ".htm", ".cshtml", ".aspx", ".ascx", ".js", ".css"}

SCRIPT_TAG_RE = re.compile(r"<script\b(?P<attributes>[^>]*)>(?P<body>.*?)</script>", re.IGNORECASE | re.DOTALL)
STYLE_TAG_RE = re.compile(r"<style\b(?P<attributes>[^>]*)>(?P<body>.*?)</style>", re.IGNORECASE | re.DOTALL)
HTML_TAG_RE = re.compile(r"<(?P<tag>[a-zA-Z][\w:-]*)(?P<attributes>\s[^<>]*?)?>", re.IGNORECASE | re.DOTALL)
ATTRIBUTE_RE = re.compile(r'(?P<name>[\w:-]+)\s*=\s*(?P<quote>["\'])(?P<value>.*?)\2', re.IGNORECASE | re.DOTALL)
EVAL_RE = re.compile(r"\beval\s*\(", re.IGNORECASE)
FUNCTION_RE = re.compile(r"\b(?:new\s+)?Function\s*\(", re.IGNORECASE)
WEBPACK_RE = re.compile(r"__webpack_require__", re.IGNORECASE)
JQUERY_READY_RE = re.compile(r"\$\(document\)\.ready\s*\([^)]*function", re.IGNORECASE)
WEBRESOURCE_RE = re.compile(r"WebResource\.axd", re.IGNORECASE)
VIEWSTATE_RE = re.compile(r"__VIEWSTATE.*script", re.IGNORECASE | re.DOTALL)
CSS_JS_URL_RE = re.compile(r'url\(\s*(?:(["\'])(javascript\s*:.*?)\1|(javascript\s*:[^)]+))\s*\)', re.IGNORECASE | re.DOTALL)


def execute_scan(options: ScanOptions) -> ScanResult:
    options.normalize()
    result = scan(options)
    apply_cleanup(result, options)
    return result


def scan(options: ScanOptions) -> ScanResult:
    options.normalize()
    scan_root = Path(options.path)
    if not scan_root.exists():
        raise FileNotFoundError(f"the scan path '{scan_root}' does not exist")

    result = ScanResult(metadata={"framework": options.framework, "cleanup": options.cleanup, "dryRun": options.dry_run})

    if scan_root.is_file():
        if scan_root.suffix.lower() in SUPPORTED_EXTENSIONS:
            result.violations.extend(_scan_file(scan_root, options))
            result.total_files_scanned = 1
        return result

    for root, dir_names, file_names in os.walk(scan_root):
        dir_names[:] = [name for name in dir_names if not _is_excluded(Path(root) / name, options)]
        for file_name in file_names:
            path = Path(root) / file_name
            if path.suffix.lower() not in SUPPORTED_EXTENSIONS:
                continue
            try:
                result.violations.extend(_scan_file(path, options))
                result.total_files_scanned += 1
            except OSError:
                result.total_files_skipped += 1
                result.skipped_paths.append(str(path))

    return result


def _scan_file(path: Path, options: ScanOptions) -> list[Violation]:
    content = path.read_text()
    suffix = path.suffix.lower()
    if suffix in {".html", ".htm", ".cshtml", ".aspx", ".ascx"}:
        return _scan_html(path, content, options)
    if suffix == ".js":
        return _scan_js(path, content, options)
    if suffix == ".css":
        return _scan_css(str(path), content, content, 0)
    return []


def _scan_html(path: Path, content: str, options: ScanOptions) -> list[Violation]:
    violations = []
    violations.extend(_scan_inline_scripts(str(path), content))
    violations.extend(_scan_inline_styles(str(path), content))
    violations.extend(_scan_html_attributes(str(path), content))
    if options.framework == FRAMEWORK_LEGACY_DOTNET:
        violations.extend(_scan_legacy(str(path), content))
    return violations


def _scan_inline_scripts(file_path: str, content: str) -> list[Violation]:
    violations = []
    for match in SCRIPT_TAG_RE.finditer(content):
        if _has_attribute(match.group("attributes") or "", "src"):
            continue
        body = match.group("body")
        if not body or not body.strip():
            continue
        violations.append(
            Violation(
                file_path=file_path,
                line_number=_line_number(content, match.start()),
                type=VIOLATION_INLINE_SCRIPT,
                content=_truncate(body, 200),
                raw_content=body,
                severity="high",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )
    return violations


def _scan_inline_styles(file_path: str, content: str) -> list[Violation]:
    violations = []
    for match in STYLE_TAG_RE.finditer(content):
        body = match.group("body")
        if not body or not body.strip():
            continue
        violations.append(
            Violation(
                file_path=file_path,
                line_number=_line_number(content, match.start()),
                type=VIOLATION_INLINE_STYLE,
                content=_truncate(body, 200),
                raw_content=body,
                severity="medium",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )
        violations.extend(_scan_css(file_path, content, body, match.start("body")))
    return violations


def _scan_html_attributes(file_path: str, content: str) -> list[Violation]:
    violations = []
    for tag_match in HTML_TAG_RE.finditer(content):
        attributes = tag_match.group("attributes") or ""
        for attribute_match in ATTRIBUTE_RE.finditer(attributes):
            absolute_index = tag_match.start("attributes") + attribute_match.start()
            attribute_name = attribute_match.group("name")
            attribute_value = attribute_match.group("value")
            raw = attribute_match.group(0)

            if attribute_name.lower().startswith("on"):
                violations.append(
                    Violation(
                        file_path=file_path,
                        line_number=_line_number(content, absolute_index),
                        type=VIOLATION_EVENT_HANDLER,
                        content=_truncate(raw, 200),
                        raw_content=attribute_value,
                        severity="high",
                        attribute_name=attribute_name,
                        original_text=raw,
                        source_index=absolute_index,
                        source_length=len(raw),
                    )
                )

            if attribute_name.lower() == "style" and attribute_value.strip():
                violations.append(
                    Violation(
                        file_path=file_path,
                        line_number=_line_number(content, absolute_index),
                        type=VIOLATION_STYLE_ATTRIBUTE,
                        content=_truncate(attribute_value, 200),
                        raw_content=attribute_value,
                        severity="medium",
                        attribute_name=attribute_name,
                        original_text=raw,
                        source_index=absolute_index,
                        source_length=len(raw),
                    )
                )
                violations.extend(_scan_css(file_path, content, attribute_value, absolute_index))

            if attribute_value.lstrip().lower().startswith("javascript:"):
                violations.append(
                    Violation(
                        file_path=file_path,
                        line_number=_line_number(content, absolute_index),
                        type=VIOLATION_JAVASCRIPT_URL,
                        content=_truncate(raw, 200),
                        raw_content=attribute_value,
                        severity="high",
                        attribute_name=attribute_name,
                        original_text=raw,
                        source_index=absolute_index,
                        source_length=len(raw),
                    )
                )
    return violations


def _scan_legacy(file_path: str, content: str) -> list[Violation]:
    violations = []
    for match in WEBRESOURCE_RE.finditer(content):
        violations.append(
            Violation(
                file_path=file_path,
                line_number=_line_number(content, match.start()),
                type=VIOLATION_LEGACY_RESOURCE,
                content="WebResource.axd reference detected",
                severity="low",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )
    for match in VIEWSTATE_RE.finditer(content):
        violations.append(
            Violation(
                file_path=file_path,
                line_number=_line_number(content, match.start()),
                type=VIOLATION_VIEWSTATE_SCRIPT,
                content="ViewState with embedded script detected",
                severity="medium",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )
    return violations


def _scan_js(path: Path, content: str, options: ScanOptions) -> list[Violation]:
    violations = []
    for match in EVAL_RE.finditer(content):
        violations.append(
            Violation(
                file_path=str(path),
                line_number=_line_number(content, match.start()),
                type=VIOLATION_DYNAMIC_INLINE,
                content="eval() usage detected",
                severity="high",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )
    for match in FUNCTION_RE.finditer(content):
        violations.append(
            Violation(
                file_path=str(path),
                line_number=_line_number(content, match.start()),
                type=VIOLATION_DYNAMIC_INLINE,
                content="Function() constructor usage detected",
                severity="high",
                original_text=match.group(0),
                source_index=match.start(),
                source_length=match.end() - match.start(),
            )
        )

    if options.framework == FRAMEWORK_JS_MODERN:
        for match in WEBPACK_RE.finditer(content):
            violations.append(
                Violation(
                    file_path=str(path),
                    line_number=_line_number(content, match.start()),
                    type=VIOLATION_DYNAMIC_INLINE,
                    content="Webpack inline pattern detected",
                    severity="low",
                    original_text=match.group(0),
                    source_index=match.start(),
                    source_length=match.end() - match.start(),
                )
            )

    if options.framework == FRAMEWORK_JS_LEGACY:
        for match in JQUERY_READY_RE.finditer(content):
            violations.append(
                Violation(
                    file_path=str(path),
                    line_number=_line_number(content, match.start()),
                    type=VIOLATION_DYNAMIC_INLINE,
                    content="jQuery.ready with inline function detected",
                    severity="medium",
                    original_text=match.group(0),
                    source_index=match.start(),
                    source_length=match.end() - match.start(),
                )
            )

    return violations


def _scan_css(file_path: str, file_content: str, css_content: str, base_index: int) -> list[Violation]:
    violations = []
    for match in CSS_JS_URL_RE.finditer(css_content):
        value = match.group(2) or match.group(3) or ""
        absolute_index = base_index + match.start()
        violations.append(
            Violation(
                file_path=file_path,
                line_number=_line_number(file_content, absolute_index),
                type=VIOLATION_JAVASCRIPT_URL,
                content=_truncate(match.group(0), 200),
                raw_content=value,
                severity="high",
                original_text=match.group(0),
                source_index=absolute_index,
                source_length=match.end() - match.start(),
            )
        )
    return violations


def apply_cleanup(result: ScanResult, options: ScanOptions) -> None:
    if options.cleanup == CLEANUP_HASH:
        _apply_hashes(result, options)
    elif options.cleanup == CLEANUP_NONCE:
        _apply_nonces(result)
    elif options.cleanup == CLEANUP_EXTERNALIZE:
        for violation in result.violations:
            if violation.type == VIOLATION_INLINE_SCRIPT:
                violation.suggested_fix = "Externalize this inline script into a .js file in the target runtime."
            elif violation.type == VIOLATION_INLINE_STYLE:
                violation.suggested_fix = "Externalize this inline style block into a .css file in the target runtime."

    for violation in result.violations:
        if violation.suggested_fix:
            continue
        violation.suggested_fix = {
            VIOLATION_EVENT_HANDLER: "Move inline event-handler code into a script file and bind it with addEventListener.",
            VIOLATION_STYLE_ATTRIBUTE: "Move inline styles into a stylesheet or use a temporary CSP nonce/hash during migration.",
            VIOLATION_JAVASCRIPT_URL: "Replace javascript: URLs with safe links, buttons, or scripted event handlers.",
            VIOLATION_DYNAMIC_INLINE: "Refactor dynamic code execution to avoid eval() and Function() where possible.",
            VIOLATION_LEGACY_RESOURCE: "Review legacy WebResource.axd usage and prefer bundled static assets when possible.",
            VIOLATION_VIEWSTATE_SCRIPT: "Review ViewState-driven script generation and move executable code into static assets.",
        }.get(violation.type, "Review and remediate this finding before enforcing a strict CSP.")


def _apply_hashes(result: ScanResult, options: ScanOptions) -> None:
    algorithm = hashlib.sha256 if options.framework == FRAMEWORK_LEGACY_DOTNET else hashlib.sha384
    algorithm_name = "sha256" if options.framework == FRAMEWORK_LEGACY_DOTNET else "sha384"
    for violation in result.violations:
        if violation.type not in {VIOLATION_INLINE_SCRIPT, VIOLATION_INLINE_STYLE}:
            continue
        source = violation.raw_content or violation.content
        digest = base64.b64encode(algorithm(source.encode("utf-8")).digest()).decode("ascii")
        violation.hash = f"{algorithm_name}-{digest}"
        violation.suggested_fix = f"Add hash to CSP: '{violation.hash}'"


def _apply_nonces(result: ScanResult) -> None:
    nonce = base64.b64encode(secrets.token_bytes(16)).decode("ascii")
    for violation in result.violations:
        if violation.type in {VIOLATION_INLINE_SCRIPT, VIOLATION_INLINE_STYLE}:
            violation.suggested_fix = f"Add nonce='{nonce}' and include 'nonce-{nonce}' in CSP"


def _is_excluded(path: Path, options: ScanOptions) -> bool:
    base = path.name.lower()
    normalized = str(path.resolve()).lower()
    for excluded in options.excluded_directories:
        cleaned = excluded.lower()
        if cleaned == base or cleaned == normalized or Path(cleaned).name == base:
            return True
    return False


def _has_attribute(attributes: str, name: str) -> bool:
    return any(match.group("name").lower() == name.lower() for match in ATTRIBUTE_RE.finditer(attributes))


def _line_number(content: str, index: int) -> int:
    return content.count("\n", 0, max(index, 0)) + 1


def _truncate(content: str, max_length: int) -> str:
    return content if len(content) <= max_length else content[:max_length] + "..."

File: go/internal/cspguard/scanner.go
package cspguard

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var (
	scriptTagPattern = regexp.MustCompile(`(?is)<script\b([^>]*)>(.*?)</script>`)
	styleTagPattern  = regexp.MustCompile(`(?is)<style\b([^>]*)>(.*?)</style>`)
	htmlTagPattern   = regexp.MustCompile(`(?is)<([a-zA-Z][\w:-]*)(\s[^<>]*?)?>`)
	evalPattern      = regexp.MustCompile(`(?i)\beval\s*\(`)
	functionPattern  = regexp.MustCompile(`(?i)\b(?:new\s+)?Function\s*\(`)
	webResource      = regexp.MustCompile(`(?i)WebResource\.axd`)
	viewState        = regexp.MustCompile(`(?is)__VIEWSTATE.*script`)
)

var supportedExtensions = map[string]struct{}{
	".html":   {},
	".htm":    {},
	".cshtml": {},
	".aspx":   {},
	".ascx":   {},
	".js":     {},
	".css":    {},
}

type invalidPathError struct {
	message string
}

func (e invalidPathError) Error() string {
	return e.message
}

func ErrInvalidPath(message string) error {
	return invalidPathError{message: message}
}

func ExecuteScan(options ScanOptions) (ScanResult, error) {
	if err := options.Normalize(); err != nil {
		return ScanResult{}, err
	}

	result, err := Scan(options)
	if err != nil {
		return ScanResult{}, err
	}

	ApplyCleanup(&result, options)

	if options.DryRun {
		fmt.Print(RenderMarkdown(result, options))
		return result, nil
	}

	reportBytes, err := RenderReport(result, options)
	if err != nil {
		return ScanResult{}, err
	}

	if err := os.WriteFile(options.ReportOutput, reportBytes, 0o644); err != nil {
		return ScanResult{}, err
	}

	return result, nil
}

func Scan(options ScanOptions) (ScanResult, error) {
	if err := options.Normalize(); err != nil {
		return ScanResult{}, err
	}

	result := ScanResult{
		Metadata: map[string]any{
			"framework": options.Framework,
			"cleanup":   options.Cleanup,
			"dryRun":    options.DryRun,
		},
	}

	info, err := os.Stat(options.Path)
	if err != nil {
		if os.IsNotExist(err) {
			return ScanResult{}, ErrInvalidPath(fmt.Sprintf("the scan path '%s' does not exist", options.Path))
		}

		return ScanResult{}, err
	}

	if !info.IsDir() {
		if _, ok := supportedExtensions[strings.ToLower(filepath.Ext(options.Path))]; !ok {
			return result, nil
		}

		violations, err := scanFile(options, options.Path)
		if err != nil {
			return ScanResult{}, err
		}

		result.TotalFilesScanned = 1
		result.Violations = append(result.Violations, violations...)
		return result, nil
	}

	err = filepath.WalkDir(options.Path, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			result.TotalFilesSkipped++
			result.SkippedPaths = append(result.SkippedPaths, path)
			return nil
		}

		if d.IsDir() {
			if path != options.Path && isExcludedDirectory(path, options) {
				return filepath.SkipDir
			}
			return nil
		}

		if _, ok := supportedExtensions[strings.ToLower(filepath.Ext(path))]; !ok {
			return nil
		}

		violations, err := scanFile(options, path)
		if err != nil {
			result.TotalFilesSkipped++
			result.SkippedPaths = append(result.SkippedPaths, path)
			return nil
		}

		result.TotalFilesScanned++
		result.Violations = append(result.Violations, violations...)
		return nil
	})
	if err != nil {
		return ScanResult{}, err
	}

	sort.Strings(result.SkippedPaths)
	return result, nil
}

func isExcludedDirectory(path string, options ScanOptions) bool {
	base := strings.ToLower(filepath.Base(path))
	for _, excluded := range options.ExcludedDirectories {
		if strings.ToLower(filepath.Base(excluded)) == base || strings.EqualFold(excluded, base) {
			return true
		}

		if filepath.IsAbs(excluded) {
			if normalized, err := filepath.Abs(path); err == nil && strings.EqualFold(normalized, excluded) {
				return true
			}
		}
	}

	return false
}

func scanFile(options ScanOptions, path string) ([]Violation, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	content := string(data)
	switch strings.ToLower(filepath.Ext(path)) {
	case ".html", ".htm", ".cshtml", ".aspx", ".ascx":
		return scanHTML(options, path, content), nil
	case ".js":
		return scanJavaScript(options, path, content), nil
	case ".css":
		return scanCSS(path, content, content, 0), nil
	default:
		return nil, nil
	}
}

func scanHTML(options ScanOptions, path, content string) []Violation {
	violations := make([]Violation, 0)
	violations = append(violations, scanInlineScripts(path, content)...)
	violations = append(violations, scanInlineStyles(path, content)...)
	violations = append(violations, scanHTMLAttributes(path, content)...)

	if options.Framework == FrameworkLegacyDotNet {
		violations = append(violations, scanLegacy(path, content)...)
	}

	return violations
}

func scanInlineScripts(path, content string) []Violation {
	violations := make([]Violation, 0)
	matches := scriptTagPattern.FindAllStringSubmatchIndex(content, -1)
	for _, match := range matches {
		fullStart, fullEnd := match[0], match[1]
		attrStart, attrEnd := match[2], match[3]
		bodyStart, bodyEnd := match[4], match[5]
		attributes := ""
		if attrStart >= 0 && attrEnd >= 0 {
			attributes = content[attrStart:attrEnd]
		}

		if hasAttribute(attributes, "src") {
			continue
		}

		body := content[bodyStart:bodyEnd]
		if strings.TrimSpace(body) == "" {
			continue
		}

		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, fullStart),
			Type:         ViolationInlineScript,
			Content:      truncateContent(body, 200),
			RawContent:   body,
			Severity:     "high",
			OriginalText: content[fullStart:fullEnd],
			SourceIndex:  fullStart,
			SourceLength: fullEnd - fullStart,
		})
	}

	return violations
}

func scanInlineStyles(path, content string) []Violation {
	violations := make([]Violation, 0)
	matches := styleTagPattern.FindAllStringSubmatchIndex(content, -1)
	for _, match := range matches {
		fullStart, fullEnd := match[0], match[1]
		bodyStart, bodyEnd := match[4], match[5]
		body := content[bodyStart:bodyEnd]
		if strings.TrimSpace(body) == "" {
			continue
		}

		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, fullStart),
			Type:         ViolationInlineStyle,
			Content:      truncateContent(body, 200),
			RawContent:   body,
			Severity:     "medium",
			OriginalText: content[fullStart:fullEnd],
			SourceIndex:  fullStart,
			SourceLength: fullEnd - fullStart,
		})

		violations = append(violations, scanCSS(path, content, body, bodyStart)...)
	}

	return violations
}

func scanHTMLAttributes(path, content string) []Violation {
	violations := make([]Violation, 0)
	matches := htmlTagPattern.FindAllStringSubmatchIndex(content, -1)
	for _, match := range matches {
		if len(match) < 6 || match[4] < 0 || match[5] < 0 {
			continue
		}

		attrStart, attrEnd := match[4], match[5]
		attributes := content[attrStart:attrEnd]
		for _, attribute := range parseAttributes(attributes) {
			absoluteIndex := attrStart + attribute.Start
			raw := attributes[attribute.Start:attribute.End]

			if strings.HasPrefix(strings.ToLower(attribute.Name), "on") {
				violations = append(violations, Violation{
					FilePath:      path,
					LineNumber:    lineNumberFromIndex(content, absoluteIndex),
					Type:          ViolationEventHandler,
					Content:       truncateContent(raw, 200),
					RawContent:    attribute.Value,
					Severity:      "high",
					AttributeName: attribute.Name,
					OriginalText:  raw,
					SourceIndex:   absoluteIndex,
					SourceLength:  attribute.End - attribute.Start,
				})
			}

			if strings.EqualFold(attribute.Name, "style") && strings.TrimSpace(attribute.Value) != "" {
				violations = append(violations, Violation{
					FilePath:      path,
					LineNumber:    lineNumberFromIndex(content, absoluteIndex),
					Type:          ViolationStyleAttribute,
					Content:       truncateContent(attribute.Value, 200),
					RawContent:    attribute.Value,
					Severity:      "medium",
					AttributeName: attribute.Name,
					OriginalText:  raw,
					SourceIndex:   absoluteIndex,
					SourceLength:  attribute.End - attribute.Start,
				})
				violations = append(violations, scanCSS(path, content, attribute.Value, absoluteIndex)...)
			}

			if strings.HasPrefix(strings.ToLower(strings.TrimSpace(attribute.Value)), "javascript:") {
				violations = append(violations, Violation{
					FilePath:      path,
					LineNumber:    lineNumberFromIndex(content, absoluteIndex),
					Type:          ViolationJavaScriptURL,
					Content:       truncateContent(raw, 200),
					RawContent:    attribute.Value,
					Severity:      "high",
					AttributeName: attribute.Name,
					OriginalText:  raw,
					SourceIndex:   absoluteIndex,
					SourceLength:  attribute.End - attribute.Start,
				})
			}
		}
	}

	return violations
}

func scanLegacy(path, content string) []Violation {
	violations := make([]Violation, 0)
	for _, match := range webResource.FindAllStringIndex(content, -1) {
		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, match[0]),
			Type:         ViolationLegacyResource,
			Content:      "WebResource.axd reference detected",
			Severity:     "low",
			OriginalText: content[match[0]:match[1]],
			SourceIndex:  match[0],
			SourceLength: match[1] - match[0],
		})
	}

	for _, match := range viewState.FindAllStringIndex(content, -1) {
		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, match[0]),
			Type:         ViolationViewStateScript,
			Content:      "ViewState with embedded script detected",
			Severity:     "medium",
			OriginalText: content[match[0]:match[1]],
			SourceIndex:  match[0],
			SourceLength: match[1] - match[0],
		})
	}

	return violations
}

func scanJavaScript(options ScanOptions, path, content string) []Violation {
	violations := make([]Violation, 0)
	for _, match := range evalPattern.FindAllStringIndex(content, -1) {
		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, match[0]),
			Type:         ViolationDynamicInline,
			Content:      "eval() usage detected",
			Severity:     "high",
			OriginalText: content[match[0]:match[1]],
			SourceIndex:  match[0],
			SourceLength: match[1] - match[0],
		})
	}

	for _, match := range functionPattern.FindAllStringIndex(content, -1) {
		violations = append(violations, Violation{
			FilePath:     path,
			LineNumber:   lineNumberFromIndex(content, match[0]),
			Type:         ViolationDynamicInline,
			Content:      "Function() constructor usage detected",
			Severity:     "high",
			OriginalText: content[match[0]:match[1]],
			SourceIndex:  match[0],
			SourceLength: match[1] - match[0],
		})
	}

	if options.Framework == FrameworkJSLegacy {
		for _, match := range regexp.MustCompile(`(?i)\$\(document\)\.ready\s*\([^)]*function`).FindAllStringIndex(content, -1) {
			violations = append(violations, Violation{
				FilePath:     path,
				LineNumber:   lineNumberFromIndex(content, match[0]),
				Type:         ViolationDynamicInline,
				Content:      "jQuery.ready with inline function detected",
				Severity:     "medium",
				OriginalText: content[match[0]:match[1]],
				SourceIndex:  match[0],
				SourceLength: match[1] - match[0],
			})
		}
	}

	if options.Framework == FrameworkJSModern {
		for _, match := range regexp.MustCompile(`(?i)__webpack_require__`).FindAllStringIndex(content, -1) {
			violations = append(violations, Violation{
				FilePath:     path,
				LineNumber:   lineNumberFromIndex(content, match[0]),
				Type:         ViolationDynamicInline,
				Content:      "Webpack inline pattern detected",
				Severity:     "low",
				OriginalText: content[match[0]:match[1]],
				SourceIndex:  match[0],
				SourceLength: match[1] - match[0],
			})
		}
	}

	return violations
}

func scanCSS(path, fileContent, cssContent string, baseIndex int) []Violation {
	violations := make([]Violation, 0)
	lowerContent := strings.ToLower(cssContent)
	searchIndex := 0

	for {
		offset := strings.Index(lowerContent[searchIndex:], "url(")
		if offset == -1 {
			break
		}

		start := searchIndex + offset
		cursor := start + 4
		for cursor < len(cssContent) && isWhitespace(cssContent[cursor]) {
			cursor++
		}

		quote := byte(0)
		if cursor < len(cssContent) && (cssContent[cursor] == '"' || cssContent[cursor] == '\'') {
			quote = cssContent[cursor]
			cursor++
		}

		valueStart := cursor
		for cursor < len(cssContent) {
			if quote != 0 {
				if cssContent[cursor] == quote {
					break
				}
			} else if cssContent[cursor] == ')' {
				break
			}
			cursor++
		}

		value := strings.TrimSpace(cssContent[valueStart:cursor])
		rawEnd := cursor
		if quote != 0 && cursor < len(cssContent) && cssContent[cursor] == quote {
			rawEnd++
			cursor++
		}
		for cursor < len(cssContent) && cssContent[cursor] != ')' {
			cursor++
		}
		if cursor < len(cssContent) && cssContent[cursor] == ')' {
			rawEnd = cursor + 1
		}

		if strings.HasPrefix(strings.ToLower(value), "javascript:") {
			absoluteStart := baseIndex + start
			violations = append(violations, Violation{
				FilePath:     path,
				LineNumber:   lineNumberFromIndex(fileContent, absoluteStart),
				Type:         ViolationJavaScriptURL,
				Content:      truncateContent(cssContent[start:rawEnd], 200),
				RawContent:   value,
				Severity:     "high",
				OriginalText: cssContent[start:rawEnd],
				SourceIndex:  absoluteStart,
				SourceLength: rawEnd - start,
			})
		}

		searchIndex = start + 4
	}

	return violations
}

func ApplyCleanup(result *ScanResult, options ScanOptions) {
	switch options.Cleanup {
	case CleanupHash:
		applyHashes(result, options)
	case CleanupNonce:
		applyNonces(result)
	case CleanupExternalize:
		applyExternalizeSuggestions(result)
	}

	applyManualSuggestions(result)
}

func applyHashes(result *ScanResult, options ScanOptions) {
	for index := range result.Violations {
		violation := &result.Violations[index]
		if violation.Type != ViolationInlineScript && violation.Type != ViolationInlineStyle {
			continue
		}

		source := violation.RawContent
		if source == "" {
			source = violation.Content
		}

		if options.Framework == FrameworkLegacyDotNet {
			hash := sha256.Sum256([]byte(source))
			violation.Hash = "sha256-" + base64.StdEncoding.EncodeToString(hash[:])
		} else {
			hash := sha512.Sum384([]byte(source))
			violation.Hash = "sha384-" + base64.StdEncoding.EncodeToString(hash[:])
		}

		violation.SuggestedFix = fmt.Sprintf("Add hash to CSP: '%s'", violation.Hash)
	}
}

func applyNonces(result *ScanResult) {
	bytes := make([]byte, 16)
	_, _ = rand.Read(bytes)
	nonce := encodeNonce(bytes)

	for index := range result.Violations {
		violation := &result.Violations[index]
		if violation.Type != ViolationInlineScript && violation.Type != ViolationInlineStyle {
			continue
		}

		violation.SuggestedFix = fmt.Sprintf("Add nonce='%s' and include 'nonce-%s' in CSP", nonce, nonce)
	}
}

func applyExternalizeSuggestions(result *ScanResult) {
	for index := range result.Violations {
		violation := &result.Violations[index]
		switch violation.Type {
		case ViolationInlineScript:
			violation.SuggestedFix = "Externalize this inline script into a .js file in the target runtime."
		case ViolationInlineStyle:
			violation.SuggestedFix = "Externalize this inline style block into a .css file in the target runtime."
		}
	}
}

func applyManualSuggestions(result *ScanResult) {
	for index := range result.Violations {
		violation := &result.Violations[index]
		if violation.SuggestedFix != "" {
			continue
		}

		switch violation.Type {
		case ViolationEventHandler:
			violation.SuggestedFix = "Move inline event-handler code into a script file and bind it with addEventListener."
		case ViolationStyleAttribute:
			violation.SuggestedFix = "Move inline styles into a stylesheet or use a temporary CSP nonce/hash during migration."
		case ViolationJavaScriptURL:
			violation.SuggestedFix = "Replace javascript: URLs with safe links, buttons, or scripted event handlers."
		case ViolationDynamicInline:
			violation.SuggestedFix = "Refactor dynamic code execution to avoid eval() and Function() where possible."
		case ViolationLegacyResource:
			violation.SuggestedFix = "Review legacy WebResource.axd usage and prefer bundled static assets when possible."
		case ViolationViewStateScript:
			violation.SuggestedFix = "Review ViewState-driven script generation and move executable code into static assets."
		default:
			violation.SuggestedFix = "Review and remediate this finding before enforcing a strict CSP."
		}
	}
}

type parsedAttribute struct {
	Name  string
	Value string
	Start int
	End   int
}

func parseAttributes(attributes string) []parsedAttribute {
	result := make([]parsedAttribute, 0)
	index := 0
	for index < len(attributes) {
		for index < len(attributes) && isWhitespace(attributes[index]) {
			index++
		}
		if index >= len(attributes) {
			break
		}

		start := index
		for index < len(attributes) && isAttributeNameChar(attributes[index]) {
			index++
		}
		if start == index {
			index++
			continue
		}

		name := attributes[start:index]
		for index < len(attributes) && isWhitespace(attributes[index]) {
			index++
		}
		if index >= len(attributes) || attributes[index] != '=' {
			continue
		}
		index++
		for index < len(attributes) && isWhitespace(attributes[index]) {
			index++
		}
		if index >= len(attributes) {
			break
		}

		valueStart := index
		if attributes[index] == '"' || attributes[index] == '\'' {
			quote := attributes[index]
			index++
			valueStart = index
			for index < len(attributes) && attributes[index] != quote {
				index++
			}
			value := attributes[valueStart:index]
			if index < len(attributes) {
				index++
			}
			result = append(result, parsedAttribute{Name: name, Value: value, Start: start, End: index})
			continue
		}

		for index < len(attributes) && !isWhitespace(attributes[index]) && attributes[index] != '>' {
			index++
		}
		value := attributes[valueStart:index]
		result = append(result, parsedAttribute{Name: name, Value: value, Start: start, End: index})
	}

	return result
}

func hasAttribute(attributes, name string) bool {
	for _, attribute := range parseAttributes(attributes) {
		if strings.EqualFold(attribute.Name, name) {
			return true
		}
	}
	return false
}

func isWhitespace(char byte) bool {
	return char == ' ' || char == '\t' || char == '\n' || char == '\r'
}

func isAttributeNameChar(char byte) bool {
	return char == ':' || char == '-' || char == '_' ||
		(char >= 'a' && char <= 'z') ||
		(char >= 'A' && char <= 'Z') ||
		(char >= '0' && char <= '9')
}

func lineNumberFromIndex(content string, index int) int {
	if index < 0 {
		return 0
	}

	line := 1
	for cursor := 0; cursor < len(content) && cursor < index; cursor++ {
		if content[cursor] == '\n' {
			line++
		}
	}

	return line
}

func truncateContent(content string, maxLength int) string {
	if len(content) <= maxLength {
		return content
	}

	return content[:maxLength] + "..."
}
